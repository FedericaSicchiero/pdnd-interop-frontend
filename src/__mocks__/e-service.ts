import { EServiceReadType } from '../../types'

export const eserviceDraft: EServiceReadType = {
  producer: {
    id: 'djofsi-sdfjdsi-djsfs',
    name: 'Comune di Bologna',
  },
  name: 'Nome e-service',
  description: 'Descrizione e-service',
  technology: 'REST',
  id: 'sdjof-sdfjdspof-dsfdsjf',
  state: 'DRAFT',
  descriptors: [],
  attributes: {
    verified: [],
    declared: [],
    certified: [
      {
        single: {
          id: 'dsdsld-dsdlds-lsdasdas',
          creationTime: '2022-02-28T16:16:18.879093Z',
          explicitAttributeVerification: false,
          verified: false,
          origin: 'dfkdsfk',
          code: 'dfjdso',
          name: 'Attributo 1',
          description: 'Descrizione attributo 1',
        },
      },
    ],
  },
}
