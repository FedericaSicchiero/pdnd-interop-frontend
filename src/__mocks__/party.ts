import { Party } from '../../types'

export const partyActiveManager: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'MANAGER',
  productInfo: { createdAt: '', id: 'interop', role: 'admin' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveDelegate: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'DELEGATE',
  productInfo: { createdAt: '', id: 'interop', role: 'admin' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveOperatorApi: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'OPERATOR',
  productInfo: { createdAt: '', id: 'interop', role: 'api' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}

export const partyActiveOperatorSecurity: Party = {
  id: 'dsofdm-dsfjds',
  attributes: [],
  state: 'ACTIVE',
  role: 'OPERATOR',
  productInfo: { createdAt: '', id: 'interop', role: 'security' },
  description: 'Comune di Test',
  institutionId: 'dsfjisdo-sdfdjs-sdfjsd',
  digitalAddress: 'comune@test.it',
}
