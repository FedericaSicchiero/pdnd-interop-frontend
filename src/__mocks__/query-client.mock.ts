import { QueryClient, QueryClientConfig } from '@tanstack/react-query'
import { queryClientConfig } from '../config/query-client'
import { deepmerge } from '@mui/utils'
import noop from 'lodash/noop'

const queryClientConfigMock: QueryClientConfig = deepmerge(
  {
    defaultOptions: {
      queries: {
        retry: false,
      },
    },
    // Disables logging
    logger: {
      log: noop,
      warn: noop,
      error: noop,
    },
  },
  queryClientConfig
)

export const queryClientMock = new QueryClient(queryClientConfigMock)
