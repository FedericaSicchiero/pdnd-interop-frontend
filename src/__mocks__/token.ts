export const token = process.env.REACT_APP_MOCK_TOKEN as string
