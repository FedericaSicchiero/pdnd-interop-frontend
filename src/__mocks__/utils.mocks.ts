import { Paginated } from './../api/comon.api.types'

/** Wraps the data passed in the pagination data structure */
export function createPaginatedMockData<T>(
  data: Array<T>,
  pagination?: Partial<Paginated<T>['pagination']>
): Paginated<T> {
  return {
    results: data,
    pagination: {
      offset: pagination?.offset ?? 0,
      limit: pagination?.offset ?? 50,
      totalResults: pagination?.offset ?? data.length,
    },
  }
}
