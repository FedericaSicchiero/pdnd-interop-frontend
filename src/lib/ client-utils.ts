import { Client } from '../../types'
import { CLIENT_COMPUTED_STATUS_LABEL } from './constants'

export function getClientComputedStatus(item: Client): keyof typeof CLIENT_COMPUTED_STATUS_LABEL {
  if (
    (item.agreement.descriptor.status === 'published' ||
      item.agreement.descriptor.status === 'deprecated') &&
    item.agreement.status === 'active' &&
    item.status === 'active'
  ) {
    return 'active'
  }

  return 'blocked'
}
