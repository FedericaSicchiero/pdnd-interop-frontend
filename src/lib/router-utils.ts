import { Location } from 'history'
import { ProviderOrSubscriber, RouteConfig } from '../../types'
import { ROUTES } from '../config/routes'
import { includesAny } from './string-utils'
import { getBits } from './url-utils'

export function isSameRoute(location: Location<unknown>, targetRoute: RouteConfig) {
  const locationBits = getBits(location)

  const sameLength = locationBits.length === targetRoute.SPLIT_PATH.length
  if (!sameLength) {
    return false
  }

  const isSamePath = targetRoute.SPLIT_PATH.every((pathFragment, i) => {
    const isDynamicFragment = pathFragment.charAt(0) === ':'
    // If this fragment of the URL is dynamic,
    // pass the check automatically, since the two will never match
    if (isDynamicFragment) {
      return true
    }

    // Otherwise, check for equality
    return pathFragment === locationBits[i]
  })

  return isSamePath
}

export function isActiveTree(location: Location<unknown>, targetRoute: RouteConfig): boolean {
  if (targetRoute.SUBROUTES) {
    const subroutes = Object.values(targetRoute.SUBROUTES)
    return Boolean(subroutes.find((r) => isActiveTree(location, r)))
  }

  return isSameRoute(location, targetRoute)
}

export function isProviderOrSubscriber(location: Location<unknown>): ProviderOrSubscriber | null {
  const locationBits = getBits(location).filter((b) => b !== 'ui')
  const mode = locationBits[0]

  if (mode === 'erogazione') {
    return 'provider'
  }

  if (mode === 'fruizione') {
    return 'subscriber'
  }

  return null
}

export function isInPlatform(location: Location<unknown>) {
  return includesAny(location.pathname, [
    ROUTES.PROVIDE.PATH,
    ROUTES.SUBSCRIBE.PATH,
    ROUTES.PROFILE.PATH,
    ROUTES.NOTIFICATION.PATH,
    ROUTES.HELP.PATH,
  ])
}
