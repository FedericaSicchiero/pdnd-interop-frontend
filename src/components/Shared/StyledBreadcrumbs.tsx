import React from 'react'
import { useLocation } from 'react-router-dom'
import { Breadcrumbs, Typography } from '@mui/material'
import { StyledLink } from './StyledLink'
import { RouteConfig } from '../../../types'
import { getBits, isSamePath } from '../../lib/router-utils'
import { ROUTES } from '../../config/routes'

export function StyledBreadcrumbs() {
  const location = useLocation()
  const currentRoute: RouteConfig | undefined = Object.values(ROUTES).find((r) =>
    isSamePath(location.pathname, r.PATH)
  )

  if (!currentRoute) {
    return null
  }

  const toDynamicPath = (route: RouteConfig) => {
    const locationBits = getBits(location)

    const dynamicSplit = route.SPLIT_PATH.map((pathFragment, i) => {
      const isDynamicFragment = pathFragment.charAt(0) === ':'
      return isDynamicFragment ? locationBits[i] : pathFragment
    })

    return `/${dynamicSplit.join('/')}`
  }

  const parentRoutes = currentRoute.PARENTS || []
  const links = [...parentRoutes, currentRoute].map((r) => ({
    label: r.LABEL,
    // Remap dynamic parts of the path to their current value
    path: toDynamicPath(r),
  }))

  // Don't display breadcrumbs for first level descentants, they are useless
  if (links.length < 2) {
    return null
  }

  return (
    <Breadcrumbs sx={{ mb: 5 }}>
      {links.map(({ label, path }, i) => {
        if (i === links.length - 1) {
          return (
            <Typography component="span" color="text.secondary" key={i}>
              {label}
            </Typography>
          )
        }
        return (
          <StyledLink key={i} to={path} sx={{ fontWeight: 700 }}>
            {label}
          </StyledLink>
        )
      })}
    </Breadcrumbs>
  )
}
