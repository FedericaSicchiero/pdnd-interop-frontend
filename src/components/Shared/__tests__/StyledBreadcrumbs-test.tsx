import React from 'react'
import renderer from 'react-test-renderer'
import { createMemoryHistory, History } from 'history'
import { AllTheProviders } from '../../../__mocks__/providers'
import { StyledBreadcrumbs } from '../StyledBreadcrumbs'
import { getDecoratedRoutes } from '../../../lib/router-utils'

function takeSnapshot(history: History<unknown>) {
  const component = renderer.create(
    <AllTheProviders defaultHistory={history}>
      <StyledBreadcrumbs />
    </AllTheProviders>
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
}

describe('Snapshot', () => {
  it('matches empty component', () => {
    const history = createMemoryHistory()
    takeSnapshot(history)
  })

  it('matches empty component', () => {
    const allRoutes = getDecoratedRoutes()
    const routes = allRoutes['it']
    const history = createMemoryHistory()
    history.push(routes.PROVIDE.PATH) // Push first level route
    takeSnapshot(history)
  })

  it('matches breadcrumbs', () => {
    const allRoutes = getDecoratedRoutes()
    const routes = allRoutes['it']
    const history = createMemoryHistory()
    history.push(routes.PROVIDE_AGREEMENT_LIST.PATH)
    takeSnapshot(history)
  })

  it('matches breadcrumbs', () => {
    const allRoutes = getDecoratedRoutes()
    const routes = allRoutes['it']
    const history = createMemoryHistory()
    history.push(routes.PROVIDE_ESERVICE_CREATE.PATH)
    takeSnapshot(history)
  })
})
