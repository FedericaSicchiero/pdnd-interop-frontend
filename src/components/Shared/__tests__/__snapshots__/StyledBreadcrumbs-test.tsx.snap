// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot matches breadcrumbs 1`] = `
<nav
  className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-igbs82-MuiTypography-root-MuiBreadcrumbs-root"
>
  <ol
    className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
  >
    <li
      className="MuiBreadcrumbs-li"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-10fszdq-MuiTypography-root-MuiLink-root"
        href="/it/erogazione"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        Erogazione
      </a>
    </li>
    <li
      aria-hidden={true}
      className="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
    >
      /
    </li>
    <li
      className="MuiBreadcrumbs-li"
    >
      <span>
        Richieste di fruizione
      </span>
    </li>
  </ol>
</nav>
`;

exports[`Snapshot matches breadcrumbs 2`] = `
<nav
  className="MuiTypography-root MuiTypography-body1 MuiBreadcrumbs-root css-igbs82-MuiTypography-root-MuiBreadcrumbs-root"
>
  <ol
    className="MuiBreadcrumbs-ol css-4pdmu4-MuiBreadcrumbs-ol"
  >
    <li
      className="MuiBreadcrumbs-li"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-10fszdq-MuiTypography-root-MuiLink-root"
        href="/it/erogazione"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        Erogazione
      </a>
    </li>
    <li
      aria-hidden={true}
      className="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
    >
      /
    </li>
    <li
      className="MuiBreadcrumbs-li"
    >
      <a
        className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-10fszdq-MuiTypography-root-MuiLink-root"
        href="/it/erogazione/e-service"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
      >
        I tuoi E-Service
      </a>
    </li>
    <li
      aria-hidden={true}
      className="MuiBreadcrumbs-separator css-1wuw8dw-MuiBreadcrumbs-separator"
    >
      /
    </li>
    <li
      className="MuiBreadcrumbs-li"
    >
      <span>
        Crea E-Service
      </span>
    </li>
  </ol>
</nav>
`;

exports[`Snapshot matches empty component 1`] = `null`;

exports[`Snapshot matches empty component 2`] = `null`;
